pkg/genericcontrolplane/aggregator.go

```diff
@@ -14,287 +14,184 @@ See the License for the specific language governing permissions and
+// Package genericcontrolplane does TODO...
 package genericcontrolplane
 
 import (
 	"fmt"
 	"net/http"
-	"strings"
-	"sync"
 
-	"k8s.io/klog/v2"
-
-	apiextensionsinformers "k8s.io/apiextensions-apiserver/pkg/client/informers/externalversions"
+	"github.com/emicklei/go-restful"
+	apiextensionsapiserver "k8s.io/apiextensions-apiserver/pkg/apiserver"
+	"k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
-	"k8s.io/apimachinery/pkg/util/sets"
-	"k8s.io/apiserver/pkg/admission"
-	genericfeatures "k8s.io/apiserver/pkg/features"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apiserver/pkg/endpoints/handlers/negotiation"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
 	genericapiserver "k8s.io/apiserver/pkg/server"
-	"k8s.io/apiserver/pkg/server/healthz"
-	genericoptions "k8s.io/apiserver/pkg/server/options"
-	utilfeature "k8s.io/apiserver/pkg/util/feature"
-	kubeexternalinformers "k8s.io/client-go/informers"
-	"k8s.io/client-go/tools/cache"
-	v1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
-	v1helper "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1/helper"
-	"k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1"
-	aggregatorapiserver "k8s.io/kube-aggregator/pkg/apiserver"
-	aggregatorscheme "k8s.io/kube-aggregator/pkg/apiserver/scheme"
-	apiregistrationclient "k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1"
-	informers "k8s.io/kube-aggregator/pkg/client/informers/externalversions/apiregistration/v1"
-	"k8s.io/kube-aggregator/pkg/controllers/autoregister"
-	"k8s.io/kubernetes/pkg/controlplane/controller/crdregistration"
-	"k8s.io/kubernetes/pkg/genericcontrolplane/options"
+	clientgoinformers "k8s.io/client-go/informers"
+	"k8s.io/klog/v2"
+	"k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator"
+	"k8s.io/kube-openapi/pkg/handler"
+	"k8s.io/kubernetes/pkg/genericcontrolplane/apis"
 )
 
-func createAggregatorConfig(
-	kubeAPIServerConfig genericapiserver.Config,
-	commandOptions *options.ServerRunOptions,
-	externalInformers kubeexternalinformers.SharedInformerFactory,
-	serviceResolver aggregatorapiserver.ServiceResolver,
-	proxyTransport *http.Transport,
-	pluginInitializers []admission.PluginInitializer,
-) (*aggregatorapiserver.Config, error) {
+var (
+	// Scheme defines methods for serializing and deserializing API objects.
+	Scheme = runtime.NewScheme()
+	// Codecs provides methods for retrieving codecs and serializers for specific
+	// versions and content types.
+	Codecs = serializer.NewCodecFactory(Scheme)
+)
+
+func init() {
+	// we need to add the options to empty v1
+	// TODO fix the server code to avoid this
+	metav1.AddToGroupVersion(Scheme, schema.GroupVersion{Version: "v1"})
+
+	// TODO: keep the generic API server from wanting this
```

Comment:
Do we need e.g. WatchEvent etc?

Reply:
This is currently only used to encode a small subset of response types. Definitely need to align on if we need a scheme here or if we should be using some other one.

Reply:
better call it discoveryScheme

Reply:
Updated